### PR TITLE
s3: Provide timestamps in the s3 file implementation

### PR DESCRIPTION
**Pull request by raphaelsc**
SSTable relies on st.st_mtime for providing creation time of data
file, which in turn is used by features like tombstone compaction.

Therefore, let's implement it.

Fixes https://github.com/scylladb/scylladb/issues/13649.

### CODE REVIEW

File: utils/s3/client.cc [hunk 1]
```diff
@@ -470,13 +470,16 @@ class client::readable_file : public file_impl {
...
+        auto object_stats = co_await _client->get_object_stats(_object_name);
         struct stat ret {};
         ret.st_nlink = 1;
         ret.st_mode = S_IFREG | S_IRUSR | S_IRGRP | S_IROTH;
-        ret.st_size = size;
+        ret.st_size = object_stats.size;
         ret.st_blksize = 1 << 10; // huh?
-        ret.st_blocks = size >> 9;
+        ret.st_blocks = object_stats.size >> 9;
+        // objects are immutable, therefore both modification and creation time can be set
```
Comment (tchaikov) [inline utils/s3/client.cc:480]: do you meant to put "neither modification nor creation time can be set"?

File: utils/s3/client.cc [hunk 2]
```diff
@@ -470,13 +470,16 @@ class client::readable_file : public file_impl {
...
-        auto size = co_await _client->get_object_size(_object_name);
+        auto object_stats = co_await _client->get_object_stats(_object_name);
         struct stat ret {};
         ret.st_nlink = 1;
         ret.st_mode = S_IFREG | S_IRUSR | S_IRGRP | S_IROTH;
-        ret.st_size = size;
+        ret.st_size = object_stats.size;
         ret.st_blksize = 1 << 10; // huh?
-        ret.st_blocks = size >> 9;
+        ret.st_blocks = object_stats.size >> 9;
```
Comment (tchaikov) [inline utils/s3/client.cc:479]: despite that this is not part of this change, this still caught my eye. if we consider 1K as the block size, then the number of blocks should be `size >> 10`. @xemul what do you think?

Comment (raphaelsc) [reply utils/s3/client.cc:479]: IIRC, st_blksize and st_blocks are unrelated. The former refers to preffered block size, for I/O efficiency. The latter applies system-specific block size. On Linux, always 512 bytes.

File: utils/s3/client.cc (hunk 1 above)
Comment (raphaelsc) [reply utils/s3/client.cc:480]: No. The reasoning is that objects are immutable on S3, therefore we can use Last-Modified to set both st_mtime and st_ctime

Comment (bhalevy) [reply utils/s3/client.cc:480]: brw, posix ctime isn't "create" time but rather, it stands for `Time of last status change` (see https://pubs.opengroup.org/onlinepubs/009696799/basedefs/sys/stat.h.html), and it is updated whenever the inode metadata changes (e.g. hardlink or chmod). Anyhow it's perfectly fine to set it equal to mtime.

File: utils/s3/client.cc (hunk 2 above)
Comment (bhalevy) [reply utils/s3/client.cc:479]: Yes, see https://pubs.opengroup.org/onlinepubs/009696799/basedefs/sys/stat.h.html
```
The unit for the st_blocks member of the stat structure is not defined within IEEE Std 1003.1-2001. In some implementations it is 512 bytes. It may differ on a file system basis. There is no correlation between values of the st_blocks and st_blksize, and the f_bsize (from [<sys/statvfs.h>](https://pubs.opengroup.org/onlinepubs/009696799/basedefs/sys/statvfs.h.html)) structure members.
```

File: utils/s3/client.cc (hunk 1 above)
Comment (raphaelsc) [reply utils/s3/client.cc:480]: > brw, posix ctime isn't "create" time but rather, it stands for `Time of last status change` (see https://pubs.opengroup.org/onlinepubs/009696799/basedefs/sys/stat.h.html), and it is updated whenever the inode metadata changes (e.g. hardlink or chmod). Anyhow it's perfectly fine to set it equal to mtime.

you're right. one refers to metadata change, whereas the other to actual data change. will update the comment. thanks.

File: utils/s3/client.cc (hunk 2 above)
Comment (tchaikov) [reply utils/s3/client.cc:479]: thanks. that makes more sense now.

File: utils/s3/client.cc (hunk 1 above)
Comment (tchaikov) [reply utils/s3/client.cc:480]: > No. The reasoning is that objects are immutable on S3, therefore we can use Last-Modified to set both st_mtime and st_ctime

I see. Could you please use this line in the comment? It helps me to understand the code better than the existing one.

File: utils/s3/client.hh [hunk 3]
```diff
@@ -32,11 +32,17 @@ class client : public enable_shared_from_this<client> {
 
 private:
     future<> get_object_header(sstring object_name, http::experimental::client::reply_handler handler);
+    std::time_t parse_http_last_modified_time(const sstring& object_name, sstring last_modified);
```
Comment (xemul) [inline utils/s3/client.hh:35]: nit: It doesn't have to be client member at all. As you noticed in the comment, it even deserves sitting in http client code
  dependent-nit: if member -- can be marked const
nit2: noexcept in any case :)

**Comment by raphaelsc** [comment]
v2: 
- make parse_http_last_modified_time() a free static function
- improve comment in readable_file::stat()

**Comment by raphaelsc** [comment]
v3: rebased in order to resolve minor conflicts
